examples/ecs/sub_states.rs

```diff
@@ -0,0 +1,272 @@
+//! This example illustrates the use of [`SubStates`] for more complex state handling patterns.
+//!
+//! [`SubStates`] are [`States`] that only exist while the App is in another [`State`]. They can
+//! be used to create more complex patterns while relying on simple enums, or to de-couple certain
+//! elements of complex state objects.
+//!
+//! In this case, we're transitioning from a `Menu` state to an `InGame` state, at which point we create
+//! a substate called `IsPaused` to track whether the game is paused or not.
+
+use bevy::prelude::*;
+
+#[derive(Debug, Clone, Copy, Default, Eq, PartialEq, Hash, States)]
+enum AppState {
+    #[default]
+    Menu,
+    InGame,
+}
+
+// In this case, instead of deriving `States`, we derive `SubStates`
+#[derive(Debug, Clone, Copy, Default, Eq, PartialEq, Hash, SubStates)]
+// And we need to add an attribute to let us know what the source state is
+// and what value it needs to have. This will ensure that unless we're
+// in [`AppState::InGame`], the [`IsPaused`] state resource
+// will not exist.
+#[source(AppState = AppState::InGame)]
```

Comment:
Non-blocking: can we reduce the syntax to just `AppState::InGame`?

Reply:
Unfortunately, due to macro's lack of ability to distinguish between type paths & enum variants, we end up needing to have something that identifies the type & then something that identifies the value... I can look into it again - but at the time I wasn't able to simplify it further.